go/pkg/diskcache/diskcache.go

```diff
@@ -0,0 +1,467 @@
+	}
+	size, err := strconv.ParseInt(pair[1], 10, 64)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid size in digest %s: %s", fname, err)
+	}
+	hash, isAc := strings.CutSuffix(pair[0], "_ac")
+	dg, err := digest.New(hash, size)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid digest from file name %s: %w", fname, err)
+	}
+	return key{digest: dg, isCas: !isAc}, nil
+}
+
+func (d *DiskCache) getPath(k key) string {
+	suffix := ""
+	if !k.isCas {
+		suffix = "_ac"
+	}
+	return filepath.Join(d.root, k.digest.Hash[:2], fmt.Sprintf("%s%s.%d", k.digest.Hash[2:], suffix, k.digest.Size))
+}
+
+func (d *DiskCache) StoreCas(dg digest.Digest, path string) error {
+	if dg.Size > int64(d.maxCapacityBytes) {
+		return fmt.Errorf("blob size %d exceeds DiskCache capacity %d", dg.Size, d.maxCapacityBytes)
+	}
+	it := &qitem{
+		key: key{digest: dg, isCas: true},
+		lat: time.Now(),
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	_, exists := d.store.LoadOrStore(it.key, it)
+	if exists {
+		return nil
+	}
+	d.mu.Lock()
+	heap.Push(d.queue, it)
+	d.mu.Unlock()
+	if err := copyFile(path, d.getPath(it.key), dg.Size); err != nil {
+		return err
+	}
+	newSize := uint64(atomic.AddInt64(&d.stats.TotalSizeBytes, dg.Size))
+	atomic.AddInt64(&d.stats.TotalNumFiles, 1)
+	atomic.AddInt64(&d.stats.NumFilesStored, 1)
+	atomic.AddInt64(&d.stats.TotalStoredBytes, dg.Size)
+	if newSize > d.maxCapacityBytes {
+		select {
+		case d.gcReq <- atomic.AddUint64(&d.gcTick, 1):
+		default:
+		}
+	}
+	return nil
+}
+
+func (d *DiskCache) StoreActionCache(dg digest.Digest, ar *repb.ActionResult) error {
+	bytes, err := proto.Marshal(ar)
+	if err != nil {
+		return err
+	}
+	size := uint64(len(bytes))
+	if size > d.maxCapacityBytes {
+		return fmt.Errorf("message size %d exceeds DiskCache capacity %d", size, d.maxCapacityBytes)
+	}
+	it := &qitem{
+		key: key{digest: dg, isCas: false},
+		lat: time.Now(),
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	d.store.Store(it.key, it)
+
+	d.mu.Lock()
+	heap.Push(d.queue, it)
+	d.mu.Unlock()
+	if err := os.WriteFile(d.getPath(it.key), bytes, 0644); err != nil {
+		return err
+	}
+	newSize := uint64(atomic.AddInt64(&d.stats.TotalSizeBytes, int64(size)))
+	atomic.AddInt64(&d.stats.TotalNumFiles, 1)
+	atomic.AddInt64(&d.stats.NumFilesStored, 1)
+	atomic.AddInt64(&d.stats.TotalStoredBytes, int64(size))
+	if newSize > d.maxCapacityBytes {
+		select {
+		case d.gcReq <- atomic.AddUint64(&d.gcTick, 1):
+		default:
+		}
+	}
+	return nil
+}
+
+func (d *DiskCache) gc() {
+	defer func() { d.gcDone <- true }()
+	for {
+		select {
+		case <-d.shutdown:
+			return
+		case <-d.ctx.Done():
+			return
+		case <-d.gcReq:
+			start := time.Now()
+			// Evict old entries until total size is below cap.
+			for uint64(atomic.LoadInt64(&d.stats.TotalSizeBytes)) > d.maxCapacityBytes {
+				d.mu.Lock()
+				it := heap.Pop(d.queue).(*qitem)
+				d.mu.Unlock()
+				size, err := d.getItemSize(it.key)
```

Comment:
I don't understand why this does a stat on the file when the size is already in the digest.